pkg/apis/networking/istio/v1alpha3/virtual_service.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2018-2019 Dynatrace LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	"bufio"
+	"bytes"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gogo/protobuf/proto"
+
+	istiov1alpha3 "istio.io/api/networking/v1alpha3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VirtualService is an Istio VirtualService resource
+type VirtualService struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec VirtualServiceSpec `json:"spec"`
+}
+
+func (vs *VirtualService) GetSpecMessage() proto.Message {
+	return &vs.Spec.VirtualService
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VirtualServiceList is a list of VirtualService resources
+type VirtualServiceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []VirtualService `json:"items"`
+}
+
+// VirtualServiceSpec is a wrapper around Istio VirtualService
+type VirtualServiceSpec struct {
+	istiov1alpha3.VirtualService
+}
+
+// DeepCopyInto is a deepcopy function, copying the receiver, writing into out. in must be non-nil.
+// Based of https://github.com/istio/istio/blob/release-0.8/pilot/pkg/config/kube/crd/types.go#L450
+func (in *VirtualServiceSpec) DeepCopyInto(out *VirtualServiceSpec) {
+	*out = *in
```

Comment:
Same here.

Reply:
This uses the deepcopy method of istio's Virtual Service, I think. Tried it with Spec and buffers. Check the tests?

Reply:
Hmm, I don't think Go can have the assignment operator overridden to use DeepCopy.

As far as I know, this instruction copies all of the fields from one object to another, including the references, but not the actual references' content. So modifying the value in some list from object, it also applies to the other. If you need actual DeepCopy you need to explicitly call something that does it.

But as I said, not sure how important is that to us.